api/src/main/java/org/openmrs/api/impl/DiagnosisServiceImpl.java

```diff
@@ -0,0 +1,208 @@
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.DiagnosisDAO;
+import org.openmrs.parameter.EncounterSearchCriteria;
+import org.openmrs.parameter.EncounterSearchCriteriaBuilder;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
+import java.util.HashSet;
+
+@Transactional
+public class DiagnosisServiceImpl extends BaseOpenmrsService implements DiagnosisService {
+	
+	private static final int PRIMARY_RANK = 1;
+	private DiagnosisDAO diagnosisDAO;
+
+	/**
+	 * Saves a diagnosis
+	 *
+	 * @param diagnosis - the diagnosis to be saved
+	 * @return the diagnosis
+	 */
+	@Override
+	public Diagnosis save(Diagnosis diagnosis) {
+		diagnosisDAO.saveDiagnosis(diagnosis);
+		return diagnosis;
+	}
+
+	/**
+	 * Voids a diagnosis
+	 *
+	 * @param diagnosis  - the diagnosis to be voided
+	 * @param voidReason - the reason for voiding the diagnosis
+	 * @return the diagnosis that was voided
+	 */
+	@Override
+	public Diagnosis voidDiagnosis(Diagnosis diagnosis, String voidReason) {
+		return Context.getDiagnosisService().save(diagnosis);
+	}
+
+	/**
+	 * Gets a diagnosis based on the uuid
+	 *
+	 * @param uuid - uuid of the diagnosis to be returned
+	 * @return diagnosis matching the given uuid
+	 */
+	@Override
+	public Diagnosis getDiagnosisByUuid(String uuid) {
+		return diagnosisDAO.getDiagnosisByUuid(uuid);
+	}
+
+	/**
+	 * Gets diagnoses since date, sorted in reverse chronological order
+	 *
+	 * @param patient  the patient whose diagnosis we are to get
+	 * @param fromDate the date used to filter diagnosis which happened from this date and later
+	 * @return the list of diagnoses for the given patient and starting from the given date
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public List<Diagnosis> getDiagnoses(Patient patient, Date fromDate) {
+		EncounterSearchCriteria encounterSearchCriteria = new EncounterSearchCriteriaBuilder().setPatient(patient)
+			.setFromDate(fromDate).createEncounterSearchCriteria();
+		List<Encounter> encounters = Context.getEncounterService().getEncounters(encounterSearchCriteria);
+		List<Diagnosis> diagnoses = new ArrayList<>();
+		for (Encounter encounter : encounters) {
+			diagnoses.addAll(encounter.getDiagnoses());
+		}
+		return diagnoses;
+	}
+
+	/**
+	 * Finds the primary diagnoses for a given encounter
+	 * The diagnosis order is identified using the integer rank value. The diagnosis rank is thus:
+	 * 1 - PRIMARY (Primary diagnosis)
+	 * 2 - SECONDARY (Secondary diagnosis)
+	 *
+	 * @param encounter the encounter whose diagnoses we are to get
+	 * @return the list of diagnoses in the given encounter whose rank is 1 (Primary diagnosis)
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public List<Diagnosis> getPrimaryDiagnoses(Encounter encounter) {
+		List<Diagnosis> diagnoses = new ArrayList<Diagnosis>();
+		if (encounter != null && !encounter.getVoided()) {
```

Comment:
Same as above

Reply:
@dkayiwa  Right here, there is no HibernateDiagnosisDAO method that handles this singly. Hence, this necessitated the looping through of diagnoses gotten and returning the primary diagnoses.

Reply:
@paradisekelechi we need to add it there. Can you ask @segunolalive to help you if you do not know how to?

Reply:
I now see why it is necessary to be there. I can and will add that right away

Reply:
Done